### PR TITLE
category

### PR DESCRIPTION
**Pull request by AlenaVyrodova**

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    @DeleteMapping("/{id}")
+    public void deleteCategory(@PathVariable Long id){
+        categoryService.deleteById(id);    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/{id}/books")
+    @Operation(summary = "Get book by category id", description = "Get book by category id")
+    public List getBooksByCategoryId(@PathVariable Long id,Pageable pageable){
+        return  bookService.findBookByCategoryId(id,pageable);
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/controller/CategoryController.java]:
```suggestion

```

File: src/main/java/com/example/bookstore/dto/book/CreateBookRequestDto.java [hunk 2]
```diff
@@ -4,22 +4,26 @@
...
     @NotNull
-    private String author;
+     String author,
     @NotNull
-    private String isbn;
+   String isbn,
     @NotNull
     @Min(0)
-    private BigDecimal price;
+     BigDecimal price,
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/dto/book/CreateBookRequestDto.java]: Poor validation

File: src/main/java/com/example/bookstore/dto/category/CreateCategoryRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookstore.dto.category;
+
+import jakarta.validation.constraints.NotNull;
+
+public record CreateCategoryRequestDto(
+        @NotNull
+        String name,
+        String description) {
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/dto/category/CreateCategoryRequestDto.java]: What about valid description?

File: src/main/java/com/example/bookstore/model/Book.java [hunk 4]
```diff
@@ -39,6 +44,15 @@ public class Book {
     private String coverImage;
 
     @Column(nullable = false)
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/model/Book.java]:
```suggestion

```

File: src/main/java/com/example/bookstore/service/CategoryService.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import jakarta.validation.Valid;
+import org.springframework.data.domain.Pageable;
+import java.util.List;
+
+public interface CategoryService {
+    List<CategoryDto> findAll(Pageable pageable);
+    CategoryDto getById(Long id);
+    CategoryDto save(@Valid CreateCategoryRequestDto categoryDto);
+    CategoryDto update(Long id, CreateCategoryRequestDto categoryRequestDto);
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/service/CategoryService.java]:
```suggestion

```

File: src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public List findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable).stream()
+                .map(categoryMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java]:
```suggestion

```

File: src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                new EntityNotFoundException("Can't find category with such  id %d"
+                        .formatted(id))));
+    }
+
+    @Override
+    public CategoryDto save(CreateCategoryRequestDto categoryRequestDto) {
+
+        return categoryMapper.toDto(categoryRepository.save(categoryMapper
+                .toCategory(categoryRequestDto)));
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java]:
```suggestion

```

File: src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    public CategoryDto getById(Long id) {
+
+        return categoryMapper.toDto(categoryRepository.findById(id).orElseThrow(() ->
+                new EntityNotFoundException("Can't find category with such  id %d"
+                        .formatted(id))));
+    }
+
+    @Override
+    public CategoryDto save(CreateCategoryRequestDto categoryRequestDto) {
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java]:
```suggestion

```

File: src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+                .toCategory(categoryRequestDto)));
+
+    }
+
+    @Override
+    public CategoryDto update(Long id, CreateCategoryRequestDto categoryRequestDto) {
+        if (!categoryRepository.existsById(id)) {
+            throw new EntityNotFoundException("There is not book in db by id %d"
+                    .formatted(id));
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/service/imp/CategoryServiceImpl.java]:
```suggestion

```

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Review by Rommelua — APPROVED** [review]

**Pull request merged**